core/gui/src/app/workspace/component/code-editor-dialog/code-editor.component.ts

```diff
@@ -169,11 +184,142 @@ export class CodeEditorComponent implements AfterViewInit, SafeStyle, OnDestroy
+
+    // Check if the AI provider is "openai"
+    this.aiAssistantService
+      .checkAIAssistantEnabled()
+      .pipe(untilDestroyed(this))
+      .subscribe({
+        next: (isEnabled: string) => {
+          if (isEnabled === "OpenAI") {
+            // "Add Type Annotation" Button
+            editor.addAction({
+              id: "type-annotation-action",
+              label: "Add Type Annotation",
+              contextMenuGroupId: "1_modification",
+              contextMenuOrder: 1.0,
+              run: ed => {
+                // User selected code (including range and content)
+                const selection = ed.getSelection();
+                const model = ed.getModel();
+                if (!model || !selection) {
+                  return;
+                }
+                // All the code in Python UDF
+                const allcode = model.getValue();
+                // Content of user selected code
+                const code = model.getValueInRange(selection);
+                // Start line of the selected code
+                const lineNumber = selection.startLineNumber;
+                this.handleTypeAnnotation(
+                  code,
+                  selection,
+                  ed as monaco.editor.IStandaloneCodeEditor,
+                  lineNumber,
+                  allcode
+                );
+              },
+            });
+          }
+        },
+      });
     if (this.language == "python") {
       this.connectLanguageServer();
     }
   }
 
+  private handleTypeAnnotation(
+    code: string,
+    range: monaco.Range,
+    editor: monaco.editor.IStandaloneCodeEditor,
+    lineNumber: number,
+    allcode: string
+  ): void {
+    this.aiAssistantService
+      .getTypeAnnotations(code, lineNumber, allcode)
+      .pipe(takeUntil(this.workflowVersionStreamSubject))
+      .subscribe({
+        next: (response: TypeAnnotationResponse) => {
+          const choices = response.choices || [];
+          if (choices.length > 0 && choices[0].message && choices[0].message.content) {
+            this.currentSuggestion = choices[0].message.content.trim();
+            this.currentCode = code;
+            this.currentRange = range;
+
+            const position = editor.getScrolledVisiblePosition(range.getStartPosition());
+            if (position) {
+              this.suggestionTop = position.top + 100;
+              this.suggestionLeft = position.left + 100;
+            }
+
+            this.showAnnotationSuggestion = true;
+
+            if (this.annotationSuggestion) {
+              this.annotationSuggestion.code = this.currentCode;
+              this.annotationSuggestion.suggestion = this.currentSuggestion;
+              this.annotationSuggestion.top = this.suggestionTop;
+              this.annotationSuggestion.left = this.suggestionLeft;
+            }
+          } else {
+            console.error("Error: OpenAI response does not contain valid message content", response);
+          }
```

Comment:
do we want users to see such an error? console.error will only be shown to developers.

Reply:
I think user don't need to see such an error.